rs/rust_canisters/random_traffic_test/src/main.rs

```diff
@@ -114,121 +117,184 @@ fn instructions_count_range() -> RangeInclusive<u32> {
-    // and `CALL_WEIGHT`.
-    //
-    // Note: Awaiting each call before possibly making another one also ensures there is only ever
-    // one response despite potentially multiple downstream calls, something that doesn't matter for
-    // the heartbeat where calls are not awaited because the heartbeat can't give a response anyway.
-    if should_reply_now() || try_call(handle_call).is_err() {
-        let payload_bytes = gen_range(reply_bytes_range());
-        let instructions_count = gen_range(instructions_count_range());
-
-        // Do some thinking.
-        let counts = api::performance_counter(0) + instructions_count as u64;
-        while counts > api::performance_counter(0) {}
-
-        let msg = vec![0_u8; payload_bytes as usize];
-        api::reply(&msg[..]);
+/// - if it tells us not to do so but the attempted downstream call fails for any reason.
+#[update]
+async fn handle_call(msg: Message) -> Vec<u8> {
+    // Samples a weighted binomial distribution to decide whether to make a downstream call (true)
+    // or reply (false). Defaults to `false` for bad weights (e.g. both 0).
+    fn should_make_downstream_call() -> bool {
+        RNG.with_borrow_mut(|rng| {
+            WeightedIndex::new([CALL_WEIGHT.get(), REPLY_WEIGHT.get()])
+                .map_or(false, |dist| dist.sample(rng) == 0)
+        })
     }
-}
 
-/// Sends out calls to `handle_call()` on random instances of this canister as provided by the
-/// `receivers` in `CONFIG`. This is done in quick succession without awaiting the outcome of these
-/// calls until either `MAX_CALLS_PER_HEARTBEAT` is reached or a synchronous rejection is observed.
-///
-/// This approach makes sense for the heartbeat because
-/// - it does not need to make a reply so it can freely make calls and then forget about them.
-/// - it allows for a quick generation of a lot of traffic.
-///
-/// Note that each such call may then spawn its own call tree with a random structure determined by
-/// the weights chosen, but unlike the heartbeat, the calls are awaited on each node in this call
-/// trees. In a sense the heartbeat may plant up to `MAX_CALLS_PER_HEARTBEAT` new call trees very
-/// quickly, but the trees then grow, wither and eventually die slowly.
-#[export_name = "canister_heartbeat"]
-fn heartbeat() {
-    for _ in 0..MAX_CALLS_PER_HEARTBEAT.get() {
-        // Passing a no-op to `try_call()` as the `on_response` reflects the' make calls and forget
-        // about them' approach described above.
-        if let Err(()) = try_call(|| {}) {
-            return;
+    // Make downstream calls until
+    // - sampling the distribution tells us to stop.
+    // - setting up a call fails.
+    // - a downstream call is rejected for any reason.
+    while should_make_downstream_call() {
+        let Some((future, record_index)) = setup_call(msg.call_tree_id, msg.call_depth + 1) else {
+            break;
+        };
```

Comment:
`setup_call()` returning an `Option` is really unexpected. I believe that's because `choose_receiver()` will return `None` iff `Config::receivers` is empty.

I would rather check against an empty `Config::receivers` in the constructor and have `choose_receiver()` unwrap. We don't need the (minor) code complexity or the resulting confusing behavior.

Reply:
Yes the reason is `choose_receiver()` but there may actually not a receiver but then it also doesn't make sense to tell the canister to make calls. But the receivers are in the config and the calls per heartbeat you set independently. My goal was just that the canister never traps.

You have a canister just replying if you don't give it any receivers, but you still call it from another canister. So it makes to allow for no receivers.

Moved the call to `choose_receiver()` outside `setup_call()`. `receiver` is now passed as an argument. Should make it clearer.